swiftpkg/internal/swiftpkg_build_files.bzl

```diff
@@ -393,8 +392,77 @@ def _clang_target_build_file(repository_ctx, pkg_ctx, target):
+    # default. We also disable the rules_swift generation of modulemaps to
+    # keep parity between consuming from Objective-C and Swift. Because of this
+    # we need to generate our own modulemap for Objective-C targets. We only do
+    # this if there isn't already a custom modulemap provided, matching the
+    # behavior of SPM (https://github.com/swiftlang/swift-package-manager/blob/4073657e12dc7a9699c08c691acdc087d66eb453/Sources/Build/BuildDescription/ClangModuleBuildDescription.swift#L175-L193).
+    #
+    # See `generate_modulemap.bzl` for details on the modulemap generation.
+    # See `//swiftpkg/tests/generate_modulemap_tests` package for a usage
+    # example.
+    if clang_src_info.modulemap_path:
+        hint_module_map = clang_src_info.modulemap_path
+    elif clang_src_info.hdrs:
+        modulemap_target_name = pkginfo_targets.modulemap_label_name(
+            bzl_target_name,
+        )
+        load_stmts.append(swiftpkg_generate_modulemap_load_stmt)
+        decls.append(
+            build_decls.new(
+                kind = swiftpkg_kinds.generate_modulemap,
+                name = modulemap_target_name,
+                attrs = {
+                    # We can't get a full picture of transitive dependencies
+                    # like rules_swift can (we would need to re-implement the
+                    # clang aspect). `use` declarations are only needed when
+                    # `-fmodules-decluse` is specified, so we are probably fine.
+                    #
+                    # Ideally long term the modulemap code in rules_swift can be
+                    # added to `objc_library` as well so we can stop generating
+                    # modulemaps entierly.
+                    "deps": [],
+                    "hdrs": clang_src_info.hdrs,
+                    "module_name": target.c99name,
+                    "visibility": ["//:__subpackages__"],
+                },
+            ),
+        )
+
+        # By including the modulemap as a dep of the parent target it gets
+        # propagated to all consumers automatically.
+        child_dep_names.append(modulemap_target_name)
+
+        hint_module_map = modulemap_target_name
+    else:
+        hint_module_map = None
+
+    # Create an interop hint so rules_swift can propagate transitive
+    # modulemaps correctly. Without this we won't get a modulemap generated
+    # for xcframework imports.
+    # `module_map` attr of `objc_library` is being removed.
+    aspect_hint_target_name = pkginfo_targets.swift_hint_label_name(
+        bzl_target_name,
+    )
+    load_stmts.append(swift_interop_hint_load_stmt)
+    decls.append(
+        build_decls.new(
+            kind = swift_kinds.interop_hint,
+            name = aspect_hint_target_name,
+            attrs = {
+                "module_map": hint_module_map,
+                "module_name": target.c99name,
+            },
+        ),
+    )
+    attrs["aspect_hints"] = [aspect_hint_target_name]
```

Comment:
We can't use `no_module` because it prevents rules_swift from generating module maps for transitive targets, which minimally breaks `apple_static_xcframework_import`.